🧑‍🌾  `rclcpp.test_memory_strategy` returns -11 in CI

### Generated by Generative AI

_No response_

### Operating System:

Linux arrch64, rhel and win

### ROS version or commit hash:

Rolling

### RMW implementation (if applicable):

_No response_

### RMW Configuration (if applicable):

_No response_

### Client library (if applicable):

_No response_

### 'ros2 doctor --report' output

_No response_

### Steps to reproduce issue

1. Run a build in any of the jobs mentioned below
2. See tests fail

### Expected behavior

Test should pass

### Actual behavior

Test is failing

### Additional information

Reference builds:
- [Rci__nightly-connext_ubuntu_noble_amd64#558](http://build.ros2.org/job/Rci__nightly-connext_ubuntu_noble_amd64/558)
- [nightly_linux-aarch64_debug#3301](https://ci.ros2.org/job/nightly_linux-aarch64_debug/3301)
- [nightly_linux-aarch64_release#3291](https://ci.ros2.org/job/nightly_linux-aarch64_release/3291)
- [nightly_linux-aarch64_repeated#3324](https://ci.ros2.org/job/nightly_linux-aarch64_repeated/3324)
- [nightly_linux-rhel_release#2421](https://ci.ros2.org/job/nightly_linux-rhel_release/2421)
- [nightly_linux-rhel_repeated#2430](https://ci.ros2.org/job/nightly_linux-rhel_repeated/2430)
- [nightly_win_rel#3582](https://ci.ros2.org/job/nightly_win_rel/3582)
- [nightly_win_rep#3885](https://ci.ros2.org/job/nightly_win_rep/3885)

Flakyness:
job_name|last_fail|first_fail|build_count|failure_count|failure_percentage
-- | -- | -- | -- | -- | --
nightly_linux-aarch64_repeated|2025-10-16|2025-10-03|14|10|71.43
nightly_linux-aarch64_release|2025-10-16|2025-10-09|16|7|43.75
nightly_linux-aarch64_debug|2025-10-16|2025-10-09|16|8|50.0
Rci__nightly-connext_ubuntu_noble_amd64|2025-10-15|2025-10-15|15|1|6.67
nightly_win_rel|2025-10-14|2025-10-14|16|1|6.25
nightly_win_rep|2025-10-11|2025-10-03|14|2|14.29
nightly_linux-rhel_repeated|2025-10-03|2025-10-03|16|1|6.25
nightly_linux-rhel_release|2025-10-03|2025-10-03|16|1|6.25

Log output:
```
-- run_test.py: invoking following command in '/home/jenkins-agent/workspace/nightly_linux-aarch64_repeated/ws/build/rclcpp/test/rclcpp':
 - /home/jenkins-agent/workspace/nightly_linux-aarch64_repeated/ws/build/rclcpp/test/rclcpp/test_memory_strategy --gtest_output=xml:/home/jenkins-agent/workspace/nightly_linux-aarch64_repeated/ws/build/rclcpp/test_results/rclcpp/test_memory_strategy.gtest.xml
Running main() from /home/jenkins-agent/workspace/nightly_linux-aarch64_repeated/ws/install/gtest_vendor/src/gtest_vendor/src/gtest_main.cc
[==========] Running 11 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 11 tests from TestMemoryStrategy
[ RUN      ] TestMemoryStrategy.construct_destruct
[       OK ] TestMemoryStrategy.construct_destruct (17 ms)
[ RUN      ] TestMemoryStrategy.get_subscription_by_handle
[ERROR] [1760595178.517516471] [ns.node]: Error in destruction of rcl service handle [~/get_type_description]: rcl node implementation is invalid, at /home/jenkins-agent/workspace/nightly_linux-aarch64_repeated/ws/src/ros2/rcl/rcl/src/rcl/node.c:387
[       OK ] TestMemoryStrategy.get_subscription_by_handle (37 ms)
[ RUN      ] TestMemoryStrategy.get_service_by_handle
-- run_test.py: return code -11
-- run_test.py: generate result file '/home/jenkins-agent/workspace/nightly_linux-aarch64_repeated/ws/build/rclcpp/test_results/rclcpp/test_memory_strategy.gtest.xml' with failed test
-- run_test.py: verify result file '/home/jenkins-agent/workspace/nightly_linux-aarch64_repeated/ws/build/rclcpp/test_results/rclcpp/test_memory_strategy.gtest.xml'
```

--- a/rclcpp/test/rclcpp/test_memory_strategy.cpp
+++ b/rclcpp/test/rclcpp/test_memory_strategy.cpp
@@ -129,6 +129,8 @@ TEST_F(TestMemoryStrategy, get_subscription_by_handle) {
     EXPECT_EQ(
       nullptr,
       memory_strategy()->get_subscription_by_handle(subscription_handle, weak_groups_to_nodes));
+    // Clear all handles to cleanup when node is destroyed
+    memory_strategy()->clear_handles();
   }  // Node goes out of scope
   EXPECT_EQ(
     nullptr,
@@ -181,6 +183,8 @@ TEST_F(TestMemoryStrategy, get_service_by_handle) {
     EXPECT_EQ(
       nullptr,
       memory_strategy()->get_service_by_handle(service_handle, weak_groups_to_nodes));
+    // Clear all handles to cleanup when node is destroyed
+    memory_strategy()->clear_handles();
   }  // Node goes out of scope
   EXPECT_EQ(
     nullptr,
@@ -228,6 +232,11 @@ TEST_F(TestMemoryStrategy, get_client_by_handle) {
     EXPECT_EQ(
       nullptr,
       memory_strategy()->get_client_by_handle(client_handle, weak_groups_to_nodes));
+    // Clear all handles to cleanup when node is destroyed
+    memory_strategy()->clear_handles();
+    // reset client handle before node destruction, otherwise it tries to access node
+    // via weak_ptr in its deleter
+    client_handle.reset();
   }  // Node goes out of scope
   EXPECT_EQ(
     nullptr,
@@ -276,6 +285,8 @@ TEST_F(TestMemoryStrategy, get_timer_by_handle) {
     EXPECT_EQ(
       nullptr,
       memory_strategy()->get_timer_by_handle(timer_handle, weak_groups_to_nodes));
+    // Clear all handles to cleanup when node is destroyed
+    memory_strategy()->clear_handles();
   }  // Node goes out of scope
   EXPECT_EQ(
     nullptr,
@@ -375,6 +386,8 @@ TEST_F(TestMemoryStrategy, get_group_by_subscription) {
     EXPECT_EQ(
       callback_group,
       memory_strategy()->get_group_by_subscription(subscription, weak_groups_to_nodes));
+    // Clear all handles to cleanup when node is destroyed
+    memory_strategy()->clear_handles();
   }  // Node goes out of scope
   // NodeBase(SubscriptionBase->rcl_node_t->NodeBase) is still alive.
   EXPECT_EQ(
@@ -420,6 +433,8 @@ TEST_F(TestMemoryStrategy, get_group_by_service) {
     EXPECT_EQ(
       nullptr,
       memory_strategy()->get_group_by_service(service, weak_groups_to_nodes));
+    // Clear all handles to cleanup when node is destroyed
+    memory_strategy()->clear_handles();
   }  // Node goes out of scope
   EXPECT_EQ(
     nullptr,
@@ -459,6 +474,8 @@ TEST_F(TestMemoryStrategy, get_group_by_client) {
     EXPECT_EQ(
       nullptr,
       memory_strategy()->get_group_by_client(client, weak_groups_to_nodes));
+    // Clear all handles to cleanup when node is destroyed
+    memory_strategy()->clear_handles();
   }  // Node goes out of scope
   EXPECT_EQ(
     nullptr,
@@ -498,6 +515,8 @@ TEST_F(TestMemoryStrategy, get_group_by_timer) {
     EXPECT_EQ(
       nullptr,
       memory_strategy()->get_group_by_timer(timer, weak_groups_to_nodes));
+    // Clear all handles to cleanup when node is destroyed
+    memory_strategy()->clear_handles();
   }  // Node goes out of scope
   EXPECT_EQ(
     nullptr,
@@ -536,6 +555,8 @@ TEST_F(TestMemoryStrategy, get_group_by_waitable) {
     EXPECT_EQ(
       nullptr,
       memory_strategy()->get_group_by_waitable(waitable, weak_groups_to_nodes));
+    // Clear all handles to cleanup when node is destroyed
+    memory_strategy()->clear_handles();
   }  // Node goes out of scope
   EXPECT_EQ(
     nullptr,

